front-end-tools/src/components/InitComponent.tsx

```diff
@@ -189,71 +193,107 @@ export default function InitComponent(props: ConnectionProps) {
+                initTemplate = displayTypeSchemaTemplate(inputParamterTypeSchemaBuffer);
 
-            if (schemaFromModule !== undefined) {
-                schema = schemaFromModule;
+                setInputParameterTemplate(initTemplate);
             }
-
-            const inputParamterTypeSchemaBuffer = getInitContractParameterSchema(
-                toBuffer(schema, 'base64'),
-                ContractName.fromString(smartContractName),
-                2
-            );
-
-            initTemplate = displayTypeSchemaTemplate(inputParamterTypeSchemaBuffer);
-
-            setInputParameterTemplate(initTemplate);
         } catch (e) {
-            if (useModuleReferenceFromStep1) {
+            if (deriveFromModuleRefernce === DO_NOT_DERIVE.value) {
                 setSchemaError(
-                    `Could not get embedded schema from the uploaded module. Uncheck "Use Module from Step 1" checkbox to manually upload a schema or uncheck "Has Input Paramter" checkbox if this entrypoint has no input parameter.  Original error: ${e}`
+                    `Could not get schema from uploaded schema. Uncheck "Has Input Paramter" checkbox if this entrypoint has no input parameter. Original error: ${e}`
                 );
             } else {
                 setSchemaError(
-                    `Could not get schema from uploaded schema. Uncheck "Has Input Paramter" checkbox if this entrypoint has no input parameter. Original error: ${e}`
+                    `Could not get embedded schema from the module. Select "${DO_NOT_DERIVE.label}" to manually upload a schema or uncheck "Has Input Paramter" checkbox if this entrypoint has no input parameter. Original error: ${e}`
                 );
             }
         }
 
-        if (initTemplate) {
-            if (inputParameterType === 'array') {
-                form.setValue('inputParameter', JSON.stringify(JSON.parse(initTemplate), undefined, 2));
-            } else if (inputParameterType === 'object') {
-                form.setValue('inputParameter', JSON.stringify(JSON.parse(initTemplate), undefined, 2));
+        if (
+            initTemplate &&
+            (inputParameterType === INPUT_PARAMETER_TYPE_ARRAY.value ||
+                inputParameterType === INPUT_PARAMETER_TYPE_OBJECT.value)
+        ) {
+            form.setValue('inputParameter', JSON.stringify(JSON.parse(initTemplate), undefined, 2));
+        }
+    }, [hasInputParameter, smartContractName, schema, inputParameterType]);
+
+    const validateModuleReference = (value: string | undefined) => {
+        if (!value) {
+            return true;
+        }
+
+        try {
+            if (!hexRegex.test(value)) {
+                return 'Invalid module reference. Not a hex string.';
             }
+
+            setModuleReference(ModuleReference.fromHexString(value));
+
+            return true;
+        } catch (e) {
+            return `Invalid module reference. Original error: ${(e as Error).message}`;
+        }
```

Comment:
Since the regex now does not ensure that the string is 64 characters, won't this also accept something like `aa`?

Is there additional validation on submission?

Reply:
It does check for the length in the function `ModuleReference.fromHexString(value)`
https://github.com/Concordium/concordium-node-sdk-js/blob/main/packages/sdk/src/types/ModuleReference.ts#L92

https://github.com/Concordium/concordium-smart-contract-tools/pull/154#discussion_r1492449853